src/benchq/resource_estimation/graph/extrapolation_estimator.py

```diff
@@ -0,0 +1,147 @@
+from dataclasses import dataclass
+from math import ceil
+from typing import List, Optional
+
+import numpy as np
+
+from ...data_structures import BasicArchitectureModel, DecoderModel
+from .graph_estimator import GraphData, GraphResourceEstimator, ResourceInfo
+
+
+@dataclass
+class ExtrapolatedGraphData(GraphData):
+    max_node_degree_r_squared: float
+    n_measurement_steps_r_squared: float
+    n_nodes_r_squared: float
+
+
+@dataclass
+class ExtrapolatedResourceInfo(ResourceInfo):
+    n_logical_qubits_r_squared: float
+    n_measurement_steps_r_squared: float
+    n_nodes_r_squared: float
+    data_used_to_extrapolate: List[ResourceInfo]
+    steps_to_extrapolate_to: int
+
+    def __repr__(self):
+        new_necessary_info = [
+            "n_logical_qubits_r_squared",
+            "n_measurement_steps_r_squared",
+            "n_nodes_r_squared",
+        ]
+        inherited_necessary_info = super().__repr__() + "\n"
+
+        return inherited_necessary_info + "\n".join(
+            f"{info}: {getattr(self, info)}" for info in new_necessary_info
+        )
+
+
+class ExtrapolationResourceEstimator(GraphResourceEstimator):
+    def __init__(
+        self,
+        hw_model: BasicArchitectureModel,
+        steps_to_extrapolate_from: List[int],
+        decoder_model: Optional[DecoderModel] = None,
+        n_measurement_steps_fit_type: str = "logarithmic",
+    ):
+        self.hw_model = hw_model
+        self.steps_to_extrapolate_from = steps_to_extrapolate_from
+        self.decoder_model = decoder_model
+        self.n_measurement_steps_fit_type = n_measurement_steps_fit_type
+
+    def _get_extrapolated_graph_data(
+        self, data: List[ResourceInfo], steps_to_extrapolate_to: int
+    ) -> ExtrapolatedGraphData:
+
+        max_node_degree, max_node_degree_r_squared = _get_linear_extrapolation(
+            self.steps_to_extrapolate_from,
+            np.array([d.n_logical_qubits for d in data]),
+            steps_to_extrapolate_to,
+        )
+        # sometimes the n_measurement_steps is logarithmic, sometimes it's linear.
+        # we need to check which one is better by inspecting the fit
+        if self.n_measurement_steps_fit_type == "logarithmic":
```

Comment:
In future we could do both at the same time and select better?
Out of scope for this PR though.

Reply:
Maybe? I think I would rather do a more complext fit model than just doing both something like a fit to a function that can grow anywhere between constant and linear. I'll have to think about it a little more.